app/scripts/lib/router.js

```diff
@@ -159,17 +159,43 @@ define(function (require, exports, module) {
     },
 
     /**
-     * Redirect the user to the best suitable OAuth flow
+     * Redirect the user to the best suitable OAuth flow.
      */
     redirectToBestOAuthChoice: function () {
-      var account = this.user.getChooserAccount();
+      var self = this;
       var route = '/oauth/signin';
 
-      if (account.isDefault()) {
-        route = '/oauth/signup';
-      }
+      var account = this.user.getChooserAccount();
 
-      return this.navigate(route, { replace: true, trigger: true });
+      // Attempt to get email address from relier
+      var email = this.broker.relier.get('email');
+
+      // Check whether email address corresponds to an existing account and
+      // navigate to correct login/signup page.
+      if (email) {
+        return account.checkAccountExistsByEmail(email)
+        .then(function (exists) {
+          if (! exists) {
+            route = '/oauth/signup';
+          }
+        }, function (err) {
+          // In case of error, choose navigation page based on
+          // whether account is a default account
+          if (account.isDefault()) {
```

Comment:
Blanket ignoring the errors seems scary, but I can also talk myself into this being the desired behavior. If the email check fails, it's not the end of the world. If the signin or signup request fails, it is. 

What's your reasoning? If we end up swallowing the error, we should document why, as well as log the error.